language-support/hs/bindings/src/DA/Ledger/Convert.hs

```diff
@@ -96,3 +99,165 @@ lowerRecordField = \case
+                   transactionEffectiveAt,
+                   transactionEvents,
+                   transactionOffset} -> do
+    -- NOTE: "<-" is used when a field is required, "let" when a field is optional
+    trid <- raiseTransactionId transactionTransactionId
+    let cid = optional (raiseCommandId transactionCommandId)
+    let wid = optional (raiseWorkflowId transactionWorkflowId)
+    leTime <- perhaps "transactionEffectiveAt" transactionEffectiveAt >>= raiseTimestamp
+    events <- raiseList raiseEvent transactionEvents
+    ofset <- raiseAbsOffset transactionOffset
+    return Transaction {trid, cid, wid, leTime, events, ofset}
+
+raiseEvent :: LL.Event -> Perhaps Event
+raiseEvent = \case
+    LL.Event{eventEvent = Nothing} -> missing "Event"
+    LL.Event(Just (LL.EventEventArchived
+                   LL.ArchivedEvent{archivedEventEventId,
+                                    archivedEventContractId,
+                                    archivedEventTemplateId,
+                                    archivedEventWitnessParties})) -> do
+        eid <- raiseEventId archivedEventEventId
+        cid <- raiseContractId archivedEventContractId
+        tid <- perhaps "archivedEventTemplateId" archivedEventTemplateId >>= raiseTemplateId
+        witness <- raiseList raiseParty archivedEventWitnessParties
+        return ArchivedEvent{eid,cid,tid,witness}
+    LL.Event(Just (LL.EventEventCreated
+                   LL.CreatedEvent{createdEventEventId,
+                                   createdEventContractId,
+                                   createdEventTemplateId,
+                                   createdEventCreateArguments,
+                                   createdEventWitnessParties})) -> do
+        eid <- raiseEventId createdEventEventId
+        cid <- raiseContractId createdEventContractId
+        tid <- perhaps "createdEventTemplateId" createdEventTemplateId >>= raiseTemplateId
+        createArgs <- perhaps "createdEventCreateArguments" createdEventCreateArguments >>= raiseRecord
+        witness <- raiseList raiseParty createdEventWitnessParties
+        return CreatedEvent{eid,cid,tid,createArgs,witness}
+
+raiseRecord :: LL.Record -> Perhaps Record
+raiseRecord = \case
+    LL.Record{recordRecordId,recordFields} -> do
+        let rid = recordRecordId >>= optional . raiseIdentifier
+        fields <- raiseList raiseRecordField recordFields
+        return Record{rid,fields}
+
+raiseRecordField :: LL.RecordField -> Perhaps RecordField
+raiseRecordField = \case
+    LL.RecordField{recordFieldLabel,recordFieldValue} -> do
+        let label = recordFieldLabel
+        value <- perhaps "recordFieldValue" recordFieldValue >>= raiseValue
+        return RecordField{label,value}
+
+-- TODO: more cases here
+raiseValue :: LL.Value -> Perhaps Value
+raiseValue = \case
+    LL.Value Nothing -> missing "Value"
+    LL.Value (Just sum) -> case sum of
+        LL.ValueSumRecord r -> (fmap VRecord . raiseRecord) r
```

Comment:
A more idiomatic way to write this is
```haskell
fmap VRecord (raiseRecord r)
```
or
```haskell
fmap VRecord $ raiseRecord r
````
or in this special case
```haskell
VRecord <$> raiseRecord r
```
@neil-da Should we add an hlint rule for the first suggestion?

Reply:
Are you saying that the fragment `(f . g) x` is not idomatic? And `f (g x)` or `f $ g x` is better?

Reply:
Yes, that is what I meant.

Reply:
https://github.com/ndmitchell/hlint/issues/644